### PR TITLE
Gv/wh/redshift/varcharmax

### PR DESCRIPTION
**Pull request by gane5hvarma**
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-server/379)

### CODE REVIEW

File: processor/integrations/integrations.go [hunk 1]
```diff
@@ -182,8 +182,15 @@ func FilterClientIntegrations(clientEvent types.SingularEventT, destNameIDMap ma
...
-func GetDestinationURL(destID string) string {
-	return fmt.Sprintf("%s/v0/%s?whSchemaVersion=%s", destTransformURL, strings.ToLower(destID), config.GetWHSchemaVersion())
+func GetDestinationURL(destType string) string {
+	if misc.Contains(misc.WarehouseDestinations, destType) {
+		if destType == "RS" {
+			return fmt.Sprintf("%s/v0/%s?whSchemaVersion=%s&RSAlterStringToText=%s", destTransformURL, strings.ToLower(destType), config.GetWHSchemaVersion(), fmt.Sprintf("%v", config.GetVarCharMaxForRS()))
+		}
+		return fmt.Sprintf("%s/v0/%s?whSchemaVersion=%s", destTransformURL, strings.ToLower(destType), config.GetWHSchemaVersion())
+	}
+	return fmt.Sprintf("%s/v0/%s", destTransformURL, strings.ToLower(destType))
```
Comment (psrikanth88) [inline processor/integrations/integrations.go:192]: init this in a variable `fmt.Sprintf("%s/v0/%s", destTransformURL, strings.ToLower(destType))` and add to it in all the if conditions

File: router/batchrouter/batchrouter.go [hunk 2]
```diff
@@ -254,6 +254,10 @@ func (brt *HandleT) postToWarehouse(batchJobs BatchJobsT, output StorageUploadOu
 		for columnName, columnType := range columns {
 			if _, ok := schemaMap[tableName][columnName]; !ok {
 				schemaMap[tableName][columnName] = columnType
+			} else {
+				if columnType == "text" {
```
Comment (psrikanth88) [inline router/batchrouter/batchrouter.go:258]: add a comment here on why we do it? and it should be set to text only if the previous value is string/text?

File: warehouse/utils/utils.go [hunk 3]
```diff
@@ -213,13 +215,18 @@ func GetSchemaDiff(currentSchema, uploadSchema map[string]map[string]string) (di
...
 			diff.ColumnMaps[tableName] = make(map[string]string)
-			for columnName, columnVal := range uploadColumnMap {
+			for columnName, columnType := range uploadColumnMap {
 				if _, ok := currentColumnsMap[columnName]; !ok {
-					diff.ColumnMaps[tableName][columnName] = columnVal
-					diff.UpdatedSchema[tableName][columnName] = columnVal
+					diff.ColumnMaps[tableName][columnName] = columnType
+					diff.UpdatedSchema[tableName][columnName] = columnType
+				} else {
+					if columnType == "text" && currentColumnsMap[columnName] == "string" {
```
Comment (psrikanth88) [inline warehouse/utils/utils.go:223]: make it else if

File: warehouse/redshift/redshift.go [hunk 4]
```diff
@@ -654,6 +685,7 @@ func loadConfig() {
 	warehouseUploadsTable = config.GetString("Warehouse.uploadsTable", "wh_uploads")
 	stagingTablePrefix = "rudder_staging_"
 	maxParallelLoads = config.GetInt("Warehouse.redshift.maxParallelLoads", 3)
+	setVarCharMax = config.GetVarCharMaxForRS()
```
Comment (psrikanth88) [inline warehouse/redshift/redshift.go:688]: directly read it from config here like others instead of a get function?

File: utils/misc/misc.go [hunk 5]
```diff
@@ -44,6 +44,8 @@ const (
 	minPostgresVersion = 100000
 )
 
+var WarehouseDestinations = []string{"RS", "BQ", "SNOWFLAKE", "POSTGRES"}
```
Comment (SumanthPuram) [inline utils/misc/misc.go:47]: Let's move this to warehouse module instead of misc?

**Pull request merged**